File: src-tauri/src/commands/contacts.rs
use super::CommandResult;
use crate::{
    AppState,
    services::contacts::{AddContactForm, ContactsService},
    utils::MapErrToString,
};
use entity::contact;
use tauri::State;

#[tauri::command(rename_all = "snake_case")]
pub async fn add_contact(
    state: State<'_, AppState>,
    contact_form: AddContactForm,
) -> CommandResult<()> {
    let contacts = state.services.get_service::<ContactsService>();

    let added_contact = contacts
        .insert_contact(contact_form)
        .await
        .map_err_to_string()?;

    log::info!("Contact created: {:?}", added_contact);

    Ok(())
}

#[tauri::command(rename_all = "snake_case")]
pub async fn get_all_contacts(state: State<'_, AppState>) -> CommandResult<Vec<contact::Model>> {
    let contacts = state.services.get_service::<ContactsService>();

    let all_contacts = contacts.get_all().await.map_err_to_string()?;
    log::debug!("Got all contacts: {:?}", all_contacts);

    Ok(all_contacts)
}

#[tauri::command(rename_all = "snake_case")]
pub async fn get_contact_with_id(
    state: State<'_, AppState>,
    id: i32,
) -> CommandResult<Option<contact::Model>> {
    let contacts = state.services.get_service::<ContactsService>();

    let contact = contacts.get_with_id(id).await.map_err_to_string()?;
    log::debug!("Got contact with id: {:?}", contact);

    Ok(contact)
}

#[tauri::command(rename_all = "snake_case")]
pub async fn get_self(state: State<'_, AppState>) -> CommandResult<contact::Model> {
    let contacts = state.services.get_service::<ContactsService>();

    let contact = contacts.get_self().await.map_err_to_string()?;

    Ok(contact.clone())
}

File: src/lib/api/contacts.ts
import { invoke } from "@tauri-apps/api/core";
import { Contact, type ContactForm } from "./models/Contact";

export async function addContact(contact_form: ContactForm) {
  await invoke("add_contact", { contact_form: contact_form });
}

export async function getSelfContact(): Promise<Contact> {
  const json = await invoke("get_self");

  const contact = new Contact().deserialize(json as any);
  return contact;
}

export async function getAllContacts(): Promise<Contact[]> {
  const json = await invoke("get_all_contacts");

  const contacts = [];
  for (const contact of json as any) {
    contacts.push(new Contact().deserialize(contact as any));
  }
  return contacts;
}
